### PR TITLE
Fix:  Avoid triggering notification if metadata changes but not the attribute value itself #3727

### PR DESCRIPTION
**Pull request by Anjali-NEC**
Fixed issue #3727

### CODE REVIEW

**Comment by fgalan** [comment]
Thanks for your contribution!

As main comment at the present moment: .test cases should be included to cover the functionality developed by this PR.

In addition, note that the PR is breaking some existing .test. That shouldn't happen, except in justified cases (in which case, please provide the proper explanation).

**Comment by Anjali-NEC** [comment]
> As main comment at the present moment: .test cases should be included to cover the functionality developed by this PR.

Added test case in https://github.com/telefonicaid/fiware-orion/pull/4090/files#diff-fbc7a73ce9da4fb932ceed7d59dfe48aba94e9aa68b2c633c41bca1e130deb2e

> In addition, note that the PR is breaking some existing .test. That shouldn't happen, except in justified cases (in which case, please provide the proper explanation).

I am looking on it.

File: test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+echo "06. Dump the accumulator, see the notification when value of attribute changes."
+echo "==============================================================================="
+accumulatorDump
+accumulatorReset
+echo
+echo
+
+
+
+echo '07. POST /v2/entities, to create Room1 with temperature and pressure.'
```
Comment (fgalan) [inline test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test:164]: Maybe the message is wrong... the entity doesn't have pressure attribute

File: test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+accumulatorStart --pretty-print
+
+--SHELL--
+
+#
+# 01. POST /v2/entities, to create Room1 with attribute and metadata.
+# 02. POST /v2/subscriptions, to create subscription with notifyOnMetadataChange=false.
```
Comment (fgalan) [inline test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test:35]:
```suggestion
# 02. POST /v2/subscriptions, to create subscription for Room 1 with notifyOnMetadataChange=false.
```

Also in echo message and expectations

File: test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+# 03. PATCH /v2/entities/Room1/attrs, to update the value of metadata.
+# 04. Dump the accumulator, see no notification when metadata changes.
+# 05. PATCH /v2/entities/Room1/attrs, to update the value of attribute.
+# 06. Dump the accumulator, see the notification when value of attribute changes.
+# 07. POST /v2/entities, to create Room2 with attribute and metadata.
+# 08. POST /v2/subscriptions, to create subscription with notifyOnMetadataChange=true.
```
Comment (fgalan) [inline test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test:41]:
```suggestion
# 08. POST /v2/subscriptions, to create subscription for Room2 with notifyOnMetadataChange=true.
```

Also in echo message and expectations

Comment (Anjali-NEC) [reply test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4090/commits/9fbcd9cec499f0dbadee3a2a7a8985f2fd3e9c2b

File: test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test (hunk 2 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4090/commits/9fbcd9cec499f0dbadee3a2a7a8985f2fd3e9c2b

File: test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test (hunk 1 above)
Comment (Anjali-NEC) [reply test/functionalTest/cases/3727_avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself/avoid_triggering_notification_if_metadata_changes_but_not_the_attribute_value_itself.test]: Fixed in https://github.com/telefonicaid/fiware-orion/pull/4090/commits/9fbcd9cec499f0dbadee3a2a7a8985f2fd3e9c2b

**Comment by fgalan** [comment]
Thanks for the recent fixes but it seems that there are .test still failing...

**Comment by fgalan** [comment]
> I have tried to fix the failing test cases but I didn't found the suitable solution. Could you please suggest what's changes needs to be done. Thanks

To do so I'd need to do an in deep analysis of the PR, probably downloading your branch in my system and test on it. I'm afraid I cannot do it in the short term...

**Comment by fgalan** [comment]
We shouldn't see this in .test files

imagen

It seems to be a wrongly solved merge conflict.

Please fix it and let's check what the next step is.

**Comment by fgalan** [comment]
After the merging of PR https://github.com/telefonicaid/fiware-orion/pull/4332 this PR needs to be upgrades with `master` branch. Otherwise, functional test will fail.

**Comment by fgalan** [comment]
I understand this PR has been closed, as issue #3737 was finally solved in [others PRs](https://github.com/telefonicaid/fiware-orion/issues/3727#issuecomment-1425996691).

@Anjali-NEC thanks anyway!

**Pull request closed without merging**